### DISCUSSION TITLE
Dynamically generated nested queries/subqueries

### DISCUSSION THREAD

**Discussion by Turun** [rusqlite/rusqlite#1378] [Questions] [answered]
EDIT: I found [`Statement.expanded_sql()`](https://docs.rs/rusqlite/latest/rusqlite/struct.Statement.html#method.expanded_sql), which I can hopefully use in conjunction with [`Statement.raw_bind_parameter`](https://docs.rs/rusqlite/latest/rusqlite/struct.Statement.html#method.raw_bind_parameter) to get the raw SQL of the prepared statement.

I'll make this post anyway, because I didn't find _anything_ on the internet regarding dynamically generated subqueries.

It's amazing that this library is prepared to let me do deranged stuff like this!

========== original text ===========

Hi,

I am writing a small [program](https://github.com/Turun/GrepolisMap) for a game where you can filter cities on a map based on several properties. Say `PlayerName` `==` `my name`. You can also combine several constraints, like `AllianceName` `==` `The badies` and `PlayerPoints` `<=` `10 000`. I convert this to SQL by constructing a query somewhat like:

```rust
sql_text = vec!["SELECT ... FROM towns ... WHERE true"];
for (index, constraint) in constraints.iter().enumerate() {
    sql_text.append(format!(
        "{constraint.table}.{constraint.column} {constraint.comparator} ?{index}"
    ));  // results in e.g. "player.name = ?1"
}
statement = self.connection.prepare(sql_text.join(" AND "));
statement.query(
    params_from_iter(
        constraints.iter().map(|constraint| constraint.value)
    )
).mapped(/*...*/);
```
To make it actually work there is a bit of hacking going on, because `constraint.value` can be a string or a number, so I need to map it to an `EitherAorB` type, for which I implemented `ToSql` by simply calling `.to_sql()` on the wrapped types. This is awesome and enables quite complex filters! 

But it is cumbersome to input a large number of filters. So I want to extend the Constraint Comparator from the basic SQL `<=`, `=`, `>=` and `<>` to include `IN` and `NOT IN`. But I don't know how to do that. I need to construct the SQL as shown in the code block above, but without handing it off to query the database. Instead I would like to get the prepared statement and put it into another prepared statement.

**Comment by gwenn** [comment]
For dynamic `IN (?, ?, ...)` clause,
there is https://docs.rs/rusqlite/latest/rusqlite/vtab/array/index.html.
Which gives `IN (SELECT value from rarray(?{index}))` instead.

**Comment by Turun** [comment, marked as answer]
In order to mark this as completed, I'll write down what I did in the end:

1. First I build a string that can be used as a format string. This is done by doing `format!("SELECT * FROM players WHERE {} {} ?{}", column_name, comparator, index)`, which results in e.g. `SELECT * FROM players WHERE name =  ?1`, or if we need nested queries by doing  `format!("SELECT * FROM players WHERE {} {} ({{{}}})", column_name, comparator, index)`, which results in e.g. `SELECT * FROM players WHERE name IN ({1})`. The different `?1` and `{1}` are required, because I want user input to be escaped, but nested queries must not be escaped. So the former is input using the `?1` syntax, while the latter `{1}` is constructed using a runtime format macro.
2. Having this string prepared with multiple `WHERE` clauses, I turn it into a prepared `rusqlite::Statement`. I can populate that prepared statement with the required user values using [`raw_bind_parameter`](https://docs.rs/rusqlite/latest/rusqlite/struct.Statement.html#method.raw_bind_parameter) , or I can call [`expanded_sql`](https://docs.rs/rusqlite/latest/rusqlite/struct.Statement.html#method.expanded_sql) on this object to get the complete raw SQL String, when it is the nested query of another SQL string (and all `?1` and `{1}` are done).
3.  Step 2 is done recursively from inner query to outer query. 

If you prefer code, here are some incomplete snippets:

```rust
    fn sql_to_bound_statement<'a>(
        &'a self: Database,
        sql_text: &str,
        constraints: &[EmptyConstraint],
        all_selections: &[EmptyTownSelection],
    ) -> anyhow::Result<Statement<'a>> {
        let mut format_mapping = HashMap::new();  // mapping what to insert where for nested queries, mapping {1}, {2}, etc to their values
        for (index, constraint) in constraints.iter().enumerate() {
            // if there is a nested query
            if constraint.referenced_selection().is_some() {
                format_mapping.insert(
                    (index + 1).to_string(),
                    constraint.get_sql_value(self, all_selections),  // recursive call, get the inner query. 
                );
            }
        }
        let sql_text = FormatArgs::new(sql_text, &format_mapping).to_string();  // use a crate to format a string at runtime
        let mut statement = self.sql_to_prepared_statement(&sql_text)?;  // turn the &str into a rusqlite::Statement

        for (index, constraint) in constraints.iter().enumerate() {
             // this constraint does not reference another query, but is instead a simple user input. e.g. player.name == "a name"
            if constraint.referenced_selection().is_none() {
                statement.raw_bind_parameter(  // insert the user input into the right place.
                    index + 1,
                    &constraint.get_sql_value(self, all_selections),  // not recursive, simply returns e.g. "some name" for the selector player.name = "some name"
                )?;
            }
        }
        Ok(statement)
    }
```

Notice that the first for loop manipulates the string with a simple format, while the latter inserts stuff into the string with the proper SQL escaped `?1` syntax. This is because we need the nested SQL queries to be unescaped SQL, but we need the user input to be escaped. 

Where `constraint.get_sql_value()` can contain: 

```rust
    fn get_sql_value(&self: EmptyConstraint, ...) -> String {
         if (/*a simple <=, <>, >=, = constraint*/) {
             return self.value // simply the user input
         }else{
            /*lots of code*/
            return db.selection_to_sql(&selection_clause, selection, all_selections).unwrap();  // The complete SQL query in string form
        }
    }
```

which in turn calls `db.sql_to_bound_statement()` and therefore recursively constructs the nested queries

```rust
    pub fn selection_to_sql(
        &self: Database,
        selection_clause: &str,
        selection: &EmptyTownSelection,
        all_selections: &[EmptyTownSelection],
    ) -> anyhow::Result<String> {
        let sql = Self::construct_sql(selection_clause, &selection.constraints, None);
        let statement = self.sql_to_bound_statement(&sql, &selection.constraints, all_selections)?;
        statement.expanded_sql().ok_or(anyhow::Error::msg(
            "Failed to convert the selection {selection} into an SQL string",
        ))
    }
```

**Reply by gwenn** [reply]
I don't understand how you can use the same method to generate the query (`constraint.get_sql_value(self, all_selections)`) and to bind the parameter (`&constraint.get_sql_value(self, all_selections)`).

**Reply by Turun** [reply]
This method just returns a string that is then inserted into the sql string. Either escaped with `?1` or not `{1}`.   
`Constraint` contains an enum that is one of `<=`, `=`, `>=`, `<>` or `IN`, `NOT IN`. No matter which comparator it is, this method returns the right hand side of that comparison. This is either whatever the user input directly for the `?1` thing, or the complete nested SQL query, which is then inserted into the parentheses `({1})`.

It's a little bit spaghetti, because the knowledge about which kind it is is required in multiple places (see the two `if constraint.referenced_selection().is_some()`/`.is_none()` in `Database::sql_to_bound_statement` and it also is required to make the format string either `?1` or `{1}`). But it works OK, so I won't refactor it for now.

(EDIT: I should add that I am totally not sure if this is the right approach. This could be way too convoluted code and I just have not seen the elegant solution yet.)

To complete the Code samples:

```rust
    /// The SQL that is inserted into the ?1 of the prepared statement or the {1} or the formated string
    pub fn get_sql_value(&self: &EmptyConstraint, db: &Database, all_selections: &[EmptyTownSelection]) -> String {
        match self.comparator {
            Comparator::LessThan
            | Comparator::Equal
            | Comparator::GreaterThan
            | Comparator::NotEqual => self.value.clone(),
            Comparator::InSelection | Comparator::NotInSelection => {
                let definitely_true = format!(
                    "SELECT {0}.{1} FROM {0}",
                    self.constraint_type.table(),
                    self.constraint_type.property()
                );
                let definitely_false = String::new();

                if self.value.is_empty() {
                    return definitely_true;
                }

                let target_selection = all_selections
                    .iter()
                    .find(|&selection| selection.name == self.value);
                if let Some(selection) = target_selection {
                    // user has typed in a valid name
                    let selection_clause = format!(
                        "{}.{}",
                        self.constraint_type.table(),
                        self.constraint_type.property()
                    );

                    // TODO error handling
                    db.selection_to_sql(&selection_clause, selection, all_selections)
                        .unwrap()
                } else {
                    // The user typed in a selection name that does not exist. If the user wants towns that are IN
                    // this imaginary selection, they'll get none. If they want all town that are NOT IN this imaginary
                    // selection they'll get all possible ones (an empty selection does not restrict the search in any way)
                    if self.comparator == Comparator::NotInSelection {
                        definitely_true
                    } else if self.comparator == Comparator::InSelection {
                        definitely_false
                    } else {
                        unreachable!(
                            "The comparator type somehow changed between the match and the if case"
                        )
                    }
                }
            }
        }
    }
```

And `Database::construct_sql()`:

```rust
    fn construct_sql<SQL>(
        selection_clause: &str,
        filter_clauses: &[SQL],
        order_clause: Option<&str>,
    ) -> String
    where
        SQL: ToSqlFragment,
    {
        // join all the tables and then append the required WHERE clauses
        let mut sql_fragments = filter_clauses
            .iter()
            .enumerate()
            .map(|(i, x)| x.to_sql_fragment(i))
            .collect();
        let mut sql_start = vec![format!(
            "SELECT {selection_clause} from \n\
                towns \n\
                LEFT JOIN islands ON (towns.island_x = islands.x AND towns.island_y = islands.y) \n\
                LEFT JOIN offsets ON (towns.slot_number = offsets.slot_number) \n\
                LEFT JOIN players ON (towns.player_id = players.player_id) \n\
                LEFT JOIN alliances ON (players.alliance_id = alliances.alliance_id) \n\
                WHERE islands.type = offsets.type",
        )];
        sql_start.append(&mut sql_fragments);
        let mut sql_text = sql_start.join(" \nAND ");
        if let Some(text) = order_clause {
            sql_text += " \n";
            sql_text += text;
        }
        sql_text
    }

```

I'll edit my above answer to be more complete. Also I pushed my changes, so if you are really really curious you can check out the [full project](https://github.com/Turun/GrepolisMap) and run the code or poke around in it. The Database stuff is all in `src/model/database.rs`, and pulls in data from `src/emptyconstraint.rs` (the method you asked about). Master/main compiles and is usable, it just has a bug that I want to fix before making another release (at which point you would not have to compile the project yourself to test it out).

**Reply by Turun** [reply]
In linear form this approach would look something like:

```rust
let constraints =vec![
    Constraint {table_column: "players.towns", comparator: "<=", value: "350"},
    Constraint {table_column: "players.rank", comparator: "<=", value: "100"},
    Constraint {table_column: "players.alliance", comparator: "IN", value: "The Badies"},
];
let sql_string = "SELECT * FROM players WHERE true";
for (index, c) in constraints.iter().enumerate() {
    if c.comparator == "IN" {
        // appends " AND players.alliance IN ({3})" to the string
        sql_string.append(format!(" AND {} {} ({{{}}})", c.table_column, c.comparator, index + 1));  
    } else {
        // appends " AND players.towns <= ?1"
        // and     " AND players.rank <= ?2"
        sql_string.append(format!(" AND {} {} ?{}", c.table_column, c.comparator, index + 1));
    }
}
// sql_string is now 
// SELECT * FROM players WHERE true AND players.towns <= ?1 AND players.rank <= ?2 AND players.alliance IN ({3})

let mapping = HashMap::new<usize, String>();
for (index, c) in constraints.iter().enumerate() {
    if c.comparator == "IN" {
        // rembers the mapping number 3 -> "The Badies"
        // The mystery method returns the sql as a string that represents the results gained from what ever defines the badies.
        // Say all players in alliance Enemy1 or Enemy2 are badies.
        mapping.insert(index + 1, c.mystery_method());
    }
}
// inserts the nested SQL for number 3.
// for this the usual formatting macro is not possible, because it has to happen at runtime (the format macro does all its things at compile time)
let sql_string = external_crate::format(sql_string, mapping);

// sql_string is now 
// SELECT * FROM players WHERE true AND players.towns <= ?1 AND players.rank <= ?2 AND players.alliance IN (SELECT players.alliance FROM players WHERE players.alliance = "Enemy1" OR players.alliance = "Enemy2")
// most importantly, it is NOT
// SELECT * FROM players WHERE true AND players.towns <= ?1 AND players.rank <= ?2 AND players.alliance IN ("SELECT players.alliance FROM players WHERE players.alliance = \\"Enemy1\\" OR players.alliance = \\"Enemy2\\"")
// which is what we would get if we were to insert the inner sql with a ?3 thing

// to bind the other parameters we need a statement
let sql = database.connection.prepare(sql_string);

for (index, c) in constraints.iter().enumerate() {
    if c.comparator == "<=" {
        // inserts the constraint value into the statement and does so safely.
        sql.raw_bind_parameter(index + 1, c.value);
    }
}

// THIS IS THE POINT WHERE mystery_method() WOULD EXTRACT THE RAW SQL

let results = sql.raw_query().mapped_rows(/*...*/);


```
